### PR TITLE
Changes to allow code to run on SparkFun OpenLog Artemis

### PR DESCRIPTION
**Pull request by PaulZC**
Hi @jaxxzer ,

We really like the BAR02 sensor and want to add support for it on [OpenLog Artemis](https://www.sparkfun.com/products/16832).
To do that, we need to remove the hardwiring to the ```Wire``` interface. This PR does that.

Please note: I haven't updated the version in library.properties. You might want to do that after merging.

Changes:
- Moved the .h and .cpp files into a **/src** folder
 - Change init to ```init(TwoWire &wirePort = Wire);```. wirePort is optional and will default to Wire so all existing code will continue to work normally
 - Add ```.begin``` which simply calls init
 - Add the first part of @NeilBaner 's PR #21 to allow the code to compile on Mbed platforms
 - Add a slightly different fix for the second part of #21 - changed the #defines to const uint8_t
 - Add keywords.txt
 - Add a second example - for the BAR02
 
I haven't included @senceryazici 's PR #16 to support the other OverSample Rates but his changes look good. You might want to merge those too.

Enjoy!

Paul

### CODE REVIEW

File: src/MS5837.cpp [hunk 1]
```diff
@@ -19,110 +19,140 @@ MS5837::MS5837() {
 	fluidDensity = 1029;
 }
 
-bool MS5837::init() {
+bool MS5837::begin(TwoWire &wirePort) {
+	return (init(wirePort));
+}
+
+bool MS5837::init(TwoWire &wirePort) {
+	_i2cPort = &wirePort; //Grab which port the user wants us to use
```
Comment (patrickelectric) [inline src/MS5837.cpp:27]: You could move _i2cPort from ptr to reference if you do the initialization of it in the constructor

File: src/MS5837.cpp [hunk 2]
```diff
@@ -19,110 +19,140 @@ MS5837::MS5837() {
...
 
-	return false; // CRC fail
+	uint8_t version = (C[0] >> 5) & 0x7F; // Extract the sensor version
+
+	if (version == 0x00) // MS5837-02BA01
+	{
+		_model = MS5837_02BA;
+		return true;
+	}
+	else if (version == 0x15) // MS5837-02BA21
```
Comment (patrickelectric) [inline src/MS5837.cpp:62]: else unnecessary after return

File: src/MS5837.cpp [hunk 3]
```diff
@@ -19,110 +19,140 @@ MS5837::MS5837() {
...
+	{
+		_model = MS5837_02BA;
+		return true;
+	}
+	else if (version == 0x1A) // MS5837-30BA26
```
Comment (patrickelectric) [inline src/MS5837.cpp:67]: You could also break it to a else case

File: src/MS5837.cpp [hunk 4]
```diff
@@ -19,110 +19,140 @@ MS5837::MS5837() {
...
 void MS5837::setFluidDensity(float density) {
 	fluidDensity = density;
 }
 
 void MS5837::read() {
 	// Request D1 conversion
-	Wire.beginTransmission(MS5837_ADDR);
-	Wire.write(MS5837_CONVERT_D1_8192);
-	Wire.endTransmission();
+	_i2cPort->beginTransmission(MS5837_ADDR);
```
Comment (patrickelectric) [inline src/MS5837.cpp:92]: Check if i2cPort is not nullptr

File: src/MS5837.h [hunk 5]
```diff
@@ -50,14 +51,16 @@ class MS5837 {
 
 	MS5837();
 
-	bool init();
+	bool init(TwoWire &wirePort = Wire);
+  bool begin(TwoWire &wirePort = Wire); // Calls init()
```
Comment (patrickelectric) [inline src/MS5837.h:55]: wrong indentation

File: src/MS5837.h [hunk 6]
```diff
@@ -50,14 +51,16 @@ class MS5837 {
...
 
 	/** Set model of MS5837 sensor. Valid options are MS5837::MS5837_30BA (default)
 	 * and MS5837::MS5837_02BA.
 	 */
 	void setModel(uint8_t model);
+	uint8_t getModel(void);
```
Comment (patrickelectric) [inline src/MS5837.h:61]: void in constructor is unnecessary in C++ code

**Review by patrickelectric — CHANGES_REQUESTED** [review]
Hi @PaulZC! 
Thanks for the great contribution, it indeed is a great thing that you are doing!
Just some small corrections and a couple of suggestions to get it merged :)

File: src/MS5837.cpp (hunk 1 above)
Comment (PaulZC) [reply src/MS5837.cpp:27]: Yes, true, but we choose not to do it that way. All the SparkFun libraries accept the wirePort as an parameter in .begin. There are some devices that can communicate over both I2C or SPI and we deal with those by overloading .begin.

File: src/MS5837.cpp (hunk 2 above)
Comment (PaulZC) [reply src/MS5837.cpp:62]: Sorry, I don't follow. This part of the code could be structured differently, a switch case might be more elegant, but the logic is sound.

File: src/MS5837.cpp (hunk 3 above)
Comment (PaulZC) [reply src/MS5837.cpp:67]: Yes, true. But the logic is sound.

File: src/MS5837.cpp (hunk 4 above)
Comment (PaulZC) [reply src/MS5837.cpp:98]: Sure. OK. Resolved here:
https://github.com/bluerobotics/BlueRobotics_MS5837_Library/pull/22/commits/2b5747d7b7bf9c39e775cab5161cc7dc878f370b#diff-e98a2dd452878f22bfa9405f77a816ec8c6add4277c3d61c18ef3ed2a98ad598R91-R96

File: src/MS5837.h (hunk 5 above)
Comment (PaulZC) [reply src/MS5837.h]: Pesky whitespace! Resolved here:
https://github.com/bluerobotics/BlueRobotics_MS5837_Library/commit/2b5747d7b7bf9c39e775cab5161cc7dc878f370b#diff-fedaeb63fea75cce65adad3c7c284f5f797befa455548788021315c68d0f423cR55

File: src/MS5837.h (hunk 6 above)
Comment (PaulZC) [reply src/MS5837.h]: OK. Resolved here:
https://github.com/bluerobotics/BlueRobotics_MS5837_Library/commit/2b5747d7b7bf9c39e775cab5161cc7dc878f370b#diff-fedaeb63fea75cce65adad3c7c284f5f797befa455548788021315c68d0f423cR61
and here:
https://github.com/bluerobotics/BlueRobotics_MS5837_Library/commit/2b5747d7b7bf9c39e775cab5161cc7dc878f370b#diff-e98a2dd452878f22bfa9405f77a816ec8c6add4277c3d61c18ef3ed2a98ad598R82

File: src/MS5837.cpp [hunk 7]
```diff
@@ -19,110 +19,140 @@ MS5837::MS5837() {
 	fluidDensity = 1029;
 }
 
-bool MS5837::init() {
+bool MS5837::begin(TwoWire &wirePort) {
+	return (init(wirePort));
```
[the remark this answers is not in the document]
Comment (PaulZC) [reply src/MS5837.cpp:23]: Hi Patrick (@patrickelectric),
Thanks for the feedback!
We don't call Wire.begin inside the library. We expect the user to call Wire.begin externally. It saves problems with (e.g.) (re)setting the clock speed. Please see this post (scroll down to the **Don’t call Wire.begin() in the library** section).
https://www.sparkfun.com/news/3245
Yes, true, you could use a class constructor argument. We choose not to do it that way.
Paul

File: src/MS5837.cpp (hunk 2 above)
Comment (PaulZC) [reply src/MS5837.cpp]: Ah, yes, you're right. I've changed this and have added more improvements in the latest commits:
- I've defined a const uint8_t for each sensor version (based on the values in the datasheets)
- I've added **MS5837_UNRECOGNISED** as a possible value for ```_model```
- ```.init``` sets ```_model``` according to the sensor version it extracts from PROM Word 0
- if the sensor version is unrecognised, ```.init``` still returns true but sets ```_model``` to **MS5837_UNRECOGNISED**
- the code is still backward-compatible. You can still call ```setModel``` to override ```_model``` if required. (E.g. if TE has used a different sensor version ID which they have forgotten to tell us about!)

The main reason I've added all of this is that we support the MS5637 on OpenLog Artemis. It has the same address as the MS5837 and has the same memory structure. It passes the same CRC test. Before I made these changes, the MS5837 was appearing as an MS5637. The MS5637 datasheet doesn't define what the 'sensor version' bits are, so I can't update the MS5637 library with the same changes... These changes mean I can test for an MS5837 first. If ```_model``` is **MS5837_UNRECOGNISED** the I know I'm almost certainly talking to a MS5637.

Anyway, I think we've ended up with a nice solution. Everything is backward-compatible. ```.init``` sets the ```_model``` automatically, but you can override it if you want to.

Cheers!

Paul

File: src/MS5837.cpp (hunk 3 above)
Comment (PaulZC) [reply src/MS5837.cpp]: Please see the comment above...

**Comment by patrickelectric** [comment]
Hi @PaulZC, it appears that your PR is failing, can you rebase it over master branch ?

**Comment by PaulZC** [comment]
Done!
Best wishes,
Paul

**Comment by PaulZC** [comment]
Hi @patrickelectric ,
Are you able to merge this PR now?
Best wishes,
Paul

**Comment by jaxxzer** [comment]
Bump @patrickelectric

**Comment by patrickelectric** [comment]
Hi @PaulZC! 
I did the final tests and everything appears to be working, to avoid more work from your part, I did a couple of corrections and organization on the commit history, they are already on our master branch.
Thanks for you contribution and sorry for the delay!

**Review by patrickelectric — APPROVED** [review]
Merged manually.

**Comment by PaulZC** [comment]
No problem @patrickelectric (& @jaxxzer) - thank you!
All the best,
Paul

**Pull request closed without merging**